### ISSUE TITLE
Enhancement: [no-shadow] add ignoreEnumKeysShadow option

### ISSUE THREAD

**Issue by Geloosa** [typescript-eslint/typescript-eslint#5385]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-shadow/

### Description

I propose `ignoreEnumKeysShadow` flag to allow enum keys to duplicate other type names. Default is `false`.

### Fail

```typescript
"@typescript-eslint/no-shadow": ["error", { ignoreEnumKeysShadow: false }]:

type Foo = number;
enum Bar {
  Foo, // disallowed enum key shadow
  Baz
}
```

### Pass

```typescript
"@typescript-eslint/no-shadow": ["error", { ignoreEnumKeysShadow: true }]:

type Foo = number;
enum Bar {
  Foo, // allowed shadow
  Baz
}

"@typescript-eslint/no-shadow": ["error", { ignoreEnumKeysShadow: false }]:

type Foo = number;
enum Bar {
  Baz,
  Qux
}
```

**Comment by bradzacher** [comment]
Why do you want such a rule?
Why are you using enum names and variable names that are the same?
What's your usecase?
More information please!

---

There is a dangerous bug vector due to the shadowing, so I don't think an option is a good idea.
I'd want a very good usecase put forward to allow it.

Esp considering your example code doesn't error by default in the rule! [playground](https://typescript-eslint.io/play#ts=4.7.2&sourceType=module&code=C4TwDgpgBAYg9nKBeKA7ArgWwEYQE4DcAUBBplAEICGeUA3kVLAgDSOVUBeRAvkUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uQAsBDACYcA7uiiJo0DtEjgwAXxDygA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by Josh-Cena** [comment]
I think the OP is mistaking type names with value names. But even judging from the rate at which reports of "false-positives" of this rule are generated, I think an option makes sense.

No real-life example comes to mind (I don't use enums personally), but shadowing enum keys doesn't look _very_ dangerous in all cases, considering most enums are literal enums—especially with the presence of rules like `prefer-literal-enum-member`. The chance that you have a _primitive_ value that you want to use as initializer but gets unintentionally shadowed is even smaller.

**Comment by bradzacher** [comment]
> judging from the rate at which reports

https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+%22no-shadow%22+is%3Aclosed+in%3Atitle+label%3A%22working+as+intended%22++enum

6 issues in 2 years [since we began reporting on this issue](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v4.0.0) isn't really that high. It just feels high because half of them were recent.

I've been meaning to add an example to the docs, just sadly haven't had time to raise a PR.

Separately could also improve the error messages for each of the cases of the rule to help explain why things are considered shadowed to help educate users.

**Comment by infacto** [comment]
I would not provide an option. Because this are just member names of the enum which are scoped by the enum name. I would treat it as bug.

**Comment by bradzacher** [comment]
@infacto it's not a bug. It's a correct report of a shadowed variable.

For example:

```ts
const A = 2;
enum Test {
  A = 1,
  B = A,
}

console.log(Test.B);
// what should be logged?
```

You might not know it - but `1` will be logged, not `2` - because the enum member `A` creates a variable within the `Test` enum's block that references the enum member.

Put another way, the above code is the same as this:
```ts
const A = 2;
enum Test {
  A = 1,
  B = Test.A,
}

console.log(Test.B);
// what should be logged?
```

Hence - the shadowing flag is completely valid and not a bug.

**Comment by infacto** [comment]
@bradzacher Ok, thanks. Good to know. I thought you have to access only with `B = Test.A`. *sigh*
Hmm 🤔 but an enum member value must be number or string. Assigning another `enum` is not allowed afaik. But idk if and how this work in eslint. It's maybe too complex. But excluding `enum` (without option) is maybe a good solution.

**Comment by bradzacher** [comment]
Specifically for names that represent `enums` it might be okay to ignore them, but there are lots of cases that we can't handle in this manner.

For example here's some convoluted but completely valid code that doesn't at all do what you'd expect unless you understand the shadowing!

```ts
class A {
  static valueOf(): string {
    return "lol this won't work";
  }
}

enum Test {
  A = 1,
  B = A.valueOf(),
}

console.log(Test.B);
// logs 1, not a string
```

So the question is - should we make the rule asymmetrical and ignore certain top-level names even though a shadowing is occurring because it would be a TS error to use the top-level name, or should we make the rule consistent and report every single case?

As you might guess - I believe that a consistent rule is much better because that way it's very predictable how the rule will behave which means it's easy for the user to understand what's happening with their code. I believe that making a rule "magic" isn't always a good thing.

**Comment by infacto** [comment]
@bradzacher I wonder I cannot reproduce this issue in [playground](https://typescript-eslint.io/play/#ts=4.7.4&sourceType=module&code=KYOwrgtgBAKsDOAXKBvAsAKClA8iYUAvFAOQ4ByAoiQDSbYwDuA9kaTAOo632wAWAJ2AFiJGAAkASpWqYAvpkyhIsFql4AxZmAFsSGnAFVJJeYowBjZiCRQAgmwBMAbiXhocJI-VZ7bAIx0vgBCbHZmGEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaZACwEMATfAd3SnwDMHIAacbSAAQBcBPABxQBjaAEt+3Qinijc3APQFi5KrQyRqZaLkjsAviD1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA).

**Comment by infacto** [comment]
@bradzacher [This](https://github.com/typescript-eslint/typescript-eslint/issues/5385#issuecomment-1225628660) is not allowed:

> Only numeric enums can have computed members, but this expression has type 'string'. If you do not need exhaustiveness checks, consider using an object literal instead.

You cannot use `A.valueOf()` as enum value.

**Comment by infacto** [comment]
Btw. only numeric enums can have computed memebers. So [this](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8BQ1oAcCuAjEBLY0EALmEftAG5ggYCmA8gGYAUAlAFzQB2GAtlrQBOiFKmiDaRDIK7QARCAD2IaEQAWuGAHdFXAORFoOwQGs5AblEBfJDaS0AHmkWDDtHr2gAVWsRGp4AF5oAEYAGlEAIWhg2AA6KhoGFlYImyA) will not work. But [this](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8BQ1oAcCuAjEBLY0EALmEftAG5ggYCmA8gGYAUAlAFzQB2GAtlrQBOiFKmiDaRDIK7QALACYA3KIC+SdUloAPNAHtBRaLR69oAFVrERqeAF5oARgA0ogELQHsAHRUaDFlZXdSA) works.

**Comment by bradzacher** [comment]
> I wonder I cannot reproduce this issue in [playground](https://typescript-eslint.io/play/#ts=4.7.4&sourceType=module&code=KYOwrgtgBAKsDOAXKBvAsAKClA8iYUAvFAOQ4ByAoiQDSbYwDuA9kaTAOo632wAWAJ2AFiJGAAkASpWqYAvpkyhIsFql4AxZmAFsSGnAFVJJeYowBjZiCRQAgmwBMAbiXhocJI-VZ7bAIx0vgBCbHZmGEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaZACwEMATfAd3SnwDMHIAacbSAAQBcBPABxQBjaAEt+3Qinijc3APQFi5KrQyRqZaLkjsAviD1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA).

@infacto because `Two` is declared *after* `Test`. The rule doesn't treat an enum declaration as a hoisted name.

> You cannot use A.valueOf() as enum value

That would be correct ***if the enum member referenced the outer class***. But because `A.valueOf()` is actually `Test.A.valueOf()` - the return type is *always* `number` - so the return type of the outer class is irrelevant!

It is a convoluted example though for sure - I was just illustrating how one might write some very confusing code due to the shadowing.
The most common cases that users run into are either an imported name or a local variable which matches the enum member name.

**Comment by infacto** [comment]
@bradzacher Interessting. Because in my project `Two` is declared after `Test` and is still reported as shadow.
![image](https://user-images.githubusercontent.com/60390085/186416358-4ca1e909-26ea-47eb-846b-032b01a5f93e.png)
lel

**Comment by bradzacher** [comment]
@infacto you might have configured [`hoist: all`](https://eslint.org/docs/latest/rules/no-shadow#hoist-all) for the rule, which will cause the rule to treat all names as hoisted names (thus ignoring the order of the code).

**Comment by infacto** [comment]
@bradzacher Maybe, it comes from a recommended in `extends`. And it seems that I cannot override it with:

```json
{
  "no-shadow": "off",
  "@typescript-eslint/no-shadow": "warn"
}
```

It still shouts. I have to set the following to mute it: 

```json
{
  "no-shadow": "off",
  "@typescript-eslint/no-shadow":  ["warn", { "hoist": "never" }]
}
```

I don't know how the config works under the hut. Maybe deep merge?

**Comment by JoshuaKGoldberg** [comment]
> I don't know how the config works under the hut. Maybe deep merge?

https://eslint.org/docs/latest/user-guide/configuring 🙂. If the ESLint docs aren't clear on this aspect of merging, I'd suggest filing an issue on them.

Closing this as it seems the discussion has petered out. Please do post back if there's some other action you think we should take. Thanks all! ✨
